### PR TITLE
feat: GS Station contract

### PR DESCRIPTION
**Pull request by sangtran-a8**

### CODE REVIEW

File: contracts/interfaces/IGachaStation.sol [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * @dev Interface of the GachaStation
+ */
+interface IGachaStation {
+    /* ============= Structs ============= */
+    struct Reward {
+        uint256 tokenId;
+        uint256 amount;
+        address tokenAddr;
+        string tokenType; // "ERC20", "ERC721", "ERC1155"
+    }
```
Comment (thanhle-ancient8) [inline contracts/interfaces/IGachaStation.sol:15]: Remove from interface

File: contracts/interfaces/IGachaStation.sol [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+    /* ============= Events ============== */
+    event OwnerOf(uint256 id, address indexed owner, Reward indexed reward);
+    event Claimed(uint256 id, address indexed owner, Reward indexed reward);
+
+    /* ============ Functions ============ */
+    function setRewardOwner(
+        address owner,
+        Reward memory reward
+    ) external returns (uint256);
```
Comment (thanhle-ancient8) [inline contracts/interfaces/IGachaStation.sol:28]: Remove from interface

File: contracts/interfaces/IGachaStation.sol [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        uint256 amount;
+        address tokenAddr;
+        string tokenType; // "ERC20", "ERC721", "ERC1155"
+    }
+
+    /* ============== Error ============== */
+    error DupplicatedClaim(uint256 id);
+
+    /* ============= Events ============== */
+    event OwnerOf(uint256 id, address indexed owner, Reward indexed reward);
```
Comment (thanhle-ancient8) [inline contracts/interfaces/IGachaStation.sol:21]: Remove from interface
Rename to `RewardSetled`

File: contracts/interfaces/IGachaStation.sol [hunk 4]
```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: MIT
```
[the remark this answers is not in the document]
Comment (sangtran-a8) [reply contracts/interfaces/IGachaStation.sol:1]: ok bro

File: test/GachaStation.spec.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+    return { station, owner, addr, others, token };
+  }
+
+  describe("#Deployment", function () {
+    it("Should deploy contract successfully & set the right owner", async function () {
+      const { station, owner } = await loadFixture(deployGachaStationFixture);
+      expect(await station.owner()).to.be.equal(owner.address);
+    });
+  });
```
[the remark this answers is not in the document]
Comment (sangtran-a8) [reply test/GachaStation.spec.ts:36]: yes, it should have more test cases

File: contracts/gacha/GachaStation.sol [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+  IERC721Receiver,
+  IERC1155Receiver
+{
+  using Address for address;
+  using SafeERC20 for IERC20;
+
+  uint256 private _nextRewardId;
+  mapping(uint256 => uint256) private _claimedBitMap;
+  mapping(uint256 id => address) private _owners;
+  mapping(uint256 id => Reward) private _rewards;
```
Comment (robertnguyen98) [inline contracts/gacha/GachaStation.sol:27]: Check this both lines. Is id redundant?

File: contracts/gacha/GachaStation.sol [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+  /* ============= Structs ============= */
+  struct Reward {
+    uint256 amount;
+    uint256 tokenId;
+    address tokenAddr;
+    bytes32 tokenType; // "ERC20", "ERC721", "ERC1155"
+  }
+
+  /* ============ Functions ============ */
+  function deposit() external payable override {
```
Comment (robertnguyen98) [inline contracts/gacha/GachaStation.sol:43]: Is deposit redundant?

File: contracts/utils/ERC20T.sol [hunk 8]
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: MIT
```
Comment (robertnguyen98) [inline contracts/utils/ERC20T.sol:1]: have existed an ERC20 contract

File: contracts/gacha/GachaStation.sol [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+      _amount,
+      ''
+    );
+    emit TokenDeposited(msg.sender, _token, _tokenId, _amount);
+  }
+
+  /**
+   * @dev Function to set reward owner
+   */
+  function setRewardOwner(
```
Comment (robertnguyen98) [inline contracts/gacha/GachaStation.sol:75]: This function does not exist in interface

File: contracts/gacha/GachaStation.sol (hunk 6 above)
Comment (sangtran-a8) [reply contracts/gacha/GachaStation.sol:27]: no bro. It should be ok

File: contracts/gacha/interfaces/IGachaStation.sol [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * @dev Interface of the GachaStation
+ */
+interface IGachaStation {
+  /* ============== Error ============== */
+  error AlreadyClaimed(uint256 id);
+  error UnsupportedTokenType();
+
+  /* ============= Events ============== */
+  event TokenDeposited(
```
[the remark this answers is not in the document]
Comment (sangtran-a8) [reply contracts/gacha/interfaces/IGachaStation.sol:14]: change to what bro???

File: contracts/utils/ERC20T.sol (hunk 8 above)
Comment (sangtran-a8) [reply contracts/utils/ERC20T.sol:1]: keep both

File: contracts/gacha/GachaStation.sol (hunk 9 above)
Comment (sangtran-a8) [reply contracts/gacha/GachaStation.sol:75]: removed by Thanh's comment

File: contracts/gacha/interfaces/IGachaStation.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (robertnguyen98) [reply contracts/gacha/interfaces/IGachaStation.sol:14]: DepositToken or TokenDepositing

[the remark this answers is not in the document]
Comment (sangtran-a8) [reply contracts/gacha/interfaces/IGachaStation.sol:14]: this event to notify token has been deposited successfully, not depositing bro

**Review by robertnguyen98 — APPROVED** [review]

File: contracts/gacha/GachaStation.sol [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+import '@openzeppelin/contracts/utils/Address.sol';
+import '@openzeppelin/contracts/access/AccessControl.sol';
+
+import './interfaces/IGachaStation.sol';
+
+contract GachaStation is
+  IGachaStation,
+  AccessControl,
+  IERC721Receiver,
+  IERC1155Receiver
```
[the remark this answers is not in the document]
Comment (sangtran-a8) [reply contracts/gacha/GachaStation.sol:19]: ok bro. but can you help me clarify why we should have it in the interface

**Review by thanhle-ancient8 — APPROVED** [review]

**Pull request merged**